src/prefect/server/orchestration/core_policy.py

```diff
@@ -631,24 +631,26 @@ async def before_transition(
         proposed_state: Optional[states.State],
         context: TaskOrchestrationContext,
     ) -> None:
-        # permit rerunning a task if the flow is retrying
-        if proposed_state.is_running() and (
-            initial_state.is_failed()
-            or initial_state.is_crashed()
-            or initial_state.is_cancelled()
+        # Only allow departure from a happily completed state if the result is not persisted
+        if (
+            initial_state.is_completed()
+            and initial_state.data
+            and getattr(initial_state.data, "type") != "unpersisted"
         ):
-            self.original_run_count = context.run.run_count
-            self.original_retry_attempt = context.run.flow_run_run_count
+            await self.reject_transition(None, "This run is already completed.")
+
+        self.original_run_count = context.run.run_count
+        self.original_retry_attempt = context.run.flow_run_run_count
+
+        # Reset run count to reset retries
+        context.run.run_count = 0
```

Comment:
Maybe we could decrease the run count by 1?

Task A(retires=100) executed 50 times and failed if Task A canceled and executed again later it will start counting from 0 again giving it the same 100 retires.

Reply:
We reset it to zero because we want to reset the task run retries during retry of a flow run. I'm not sure we should complicate it by decrementing the count in other cases. Retries are currently the amount of times a RUNNING -> FAILED proposal should be repeated. If you go from a terminal state back to RUNNING it feels like retries should be reset as this is a "new attempt" to orchestrate the task.